### PR TITLE
feat: impl Storable for Option<T: Storable>

### PR DESCRIPTION
**Pull request by witter-deland**

### CODE REVIEW

File: src/storable/tests.rs [hunk 1]
```diff
@@ -35,4 +35,29 @@ proptest! {
     fn f32_roundtrip(v in any::<f32>()) {
         prop_assert_eq!(v, Storable::from_bytes(v.to_bytes()));
     }
+
+    #[test]
+    fn optional_f64_roundtrip(v in any::<f64>()) {
+        let opt_v= Some(v);
+        prop_assert_eq!(Some(v), Storable::from_bytes(opt_v.to_bytes()));
+    }
```
Comment (roman-kashitsyn) [inline src/storable/tests.rs:43]: We never exercise the `None` case in these tests. Let's check this path as well:

```suggestion
    fn optional_f64_roundtrip(v in proptest::option::of(any::<f64>())) {
        prop_assert_eq!(v, Storable::from_bytes(v.to_bytes()));
    }
```

File: src/storable/tests.rs [hunk 2]
```diff
@@ -35,4 +35,27 @@ proptest! {
...
+
+    #[test]
+    fn optional_string_roundtrip(v in proptest::option::of(any::<String>())) {
+        prop_assert_eq!(v.clone(), Storable::from_bytes(v.to_bytes()));
+    }
+
+    #[test]
+    fn optional_tuple_roundtrip(x in proptest::option::of(any::<u64>()), y in uniform20(any::<u8>())) {
+        let opt_tuple= if x.is_some() { Some((x.unwrap(),y))} else { None };
+        prop_assert_eq!(opt_tuple, Storable::from_bytes(opt_tuple.to_bytes()));
```
Comment (roman-kashitsyn) [inline src/storable/tests.rs:52]:
```suggestion
    fn optional_tuple_roundtrip(v in proptest::option::of((any::<u64>(), uniform20(any::<u8>())))) {
        prop_assert_eq!(v, Storable::from_bytes(v.to_bytes()));
```

File: src/storable/tests.rs [hunk 3]
```diff
@@ -35,4 +35,27 @@ proptest! {
...
+    }
+
+    #[test]
+    fn optional_tuple_variable_width_u8_roundtrip(x in  proptest::option::of(any::<u64>()), v in pvec(any::<u8>(), 0..40)) {
+        let bytes = Blob::<48>::try_from(&v[..]).unwrap();
+        let opt_tuple= if x.is_some() { Some((x.unwrap(),bytes))} else { None };
+        prop_assert_eq!(opt_tuple, Storable::from_bytes(opt_tuple.to_bytes()));
```
Comment (roman-kashitsyn) [inline src/storable/tests.rs:59]:
```suggestion
    fn optional_tuple_variable_width_u8_roundtrip(v in proptest::option::of((any::<u64>(), pvec(any::<u8>(), 0..40))) {
        let v = v.map(|(n, bytes)| (n, Blob::<48>::try_from(&bytes[..]).unwrap()));
        prop_assert_eq!(v, Storable::from_bytes(v.to_bytes()));
```

File: src/storable.rs [hunk 4]
```diff
@@ -421,6 +421,40 @@ where
     };
 }
 
+impl<T: Storable> Storable for Option<T> {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        match self {
+            Some(t) => {
+                let mut bytes = t.to_bytes().into_owned();
+                bytes.insert(0, 1);
```
Comment (roman-kashitsyn) [inline src/storable.rs:429]: Inserting the marker at the front can be inefficient for large values. Should we insert the marker at the end instead?

File: src/storable.rs [hunk 5]
```diff
@@ -421,6 +421,44 @@ where
...
+
+    fn from_bytes(bytes: Cow<[u8]>) -> Self {
+        match bytes.split_last() {
+            Some((&last, bytes)) => match last == 0 {
+                true => None,
+                false => Some(T::from_bytes(Cow::Borrowed(bytes))),
+            },
+            None => {
+                panic!("Invalid Option encoding");
+            }
```
Comment (roman-kashitsyn) [inline src/storable.rs:444]: Let's be stricter in the decoding to help users discover their errors as soon as possible. For example, we shouldn't accept `[1, 0]` as `None` and shouldn't accept byte sequences that don't end by `{0, 1}` (e.g., `[0, 3]`).
```suggestion
            Some((last, rest)) => match last {
                0 => {
                    assert!(rest.is_empty(), "Invalid Option encoding: unexpected prefix before the None marker: {rest:?}");
                    None
                }
                1 => Some(T::from_bytes(Cow::Borrowed(rest))),
                _ => panic!("Invalid Option encoding: unexpected variant marker {last}"),
            },
            None => panic!("Invalid Option encoding: expected at least one byte"),

```

**Review by roman-kashitsyn — APPROVED** [review]

**Comment by ielashi** [comment]
@witter-deland Thanks for your contribution! I believe you need to merge the latest `main` into this PR for it to be merge-able.

**Pull request merged**